src/IO/WriteBufferFromS3TaskTracker.h

```diff
@@ -20,22 +22,33 @@ class WriteBufferFromS3::TaskTracker
 public:
     using Callback = std::function<void()>;
 
-    explicit TaskTracker(ThreadPoolCallbackRunner<void> scheduler_);
+    explicit TaskTracker(ThreadPoolCallbackRunner<void> scheduler_, size_t max_tasks_inflight_ = 0);
     ~TaskTracker();
 
     static ThreadPoolCallbackRunner<void> syncRunner();
 
     bool isAsync() const;
-    void waitReady();
+    size_t consumeReady();
+    void waitAny();
     void waitAll();
     void safeWaitAll();
     void add(Callback && func);
 
 private:
-    bool is_async;
+    void waitInFlight();
+
+    const bool is_async;
     ThreadPoolCallbackRunner<void> scheduler;
-    std::list<std::future<void>> futures;
+    const size_t max_tasks_inflight;
+
+    using FutureList = std::list<std::future<void>>;
+    FutureList futures;
     Poco::Logger * log = &Poco::Logger::get("TaskTracker");
+
+    std::mutex mutex;
+    std::condition_variable cond_var;
+    using FinishedList = std::list<FutureList::iterator>;
+    FinishedList finished_futures;
```

Comment:
Actually we can.
Just wait notification with cond_var and after it iterate by all features and ask them `wait_for(0) == std::future_status::ready`

Reply:
I have a better answer again.

`wait_for(0)` is not helping when you need `waitAny`. It was useful only for collecting already finished tasks, like `waitIfAny` or `waitReady` in master. 

The code like this
```
cond_var.wait(lock, [&] () { return !<someone_finished>; });
for (auto & it : futures)
   if (it.wait_for(std::chrono::seconds(0)) == std::future_status::ready)
      it.get()
```
won't work due to race.  There is a chance that such code call no `get()` at all. Even when !<someone_finished> is true, the future might be not ready. Because the instruction `future->set()` follows after `cond_var.notify_one()` and it is out of critical scope.